pkg/inclusion/paths_test.go

```diff
@@ -338,97 +340,144 @@ func Test_genSubTreeRootPath(t *testing.T) {
-			{instructions: []WalkInstruction{WalkRight, WalkLeft, WalkLeft, WalkLeft, WalkLeft, WalkLeft}, row: 63},
-		}},
-		// last 63 shares in the last row of a 64 x 64 square.
-		{64, 4032, 63, []path{
-			{instructions: []WalkInstruction{WalkLeft, WalkLeft, WalkLeft}, row: 63},
-			{instructions: []WalkInstruction{WalkLeft, WalkLeft, WalkRight}, row: 63},
-			{instructions: []WalkInstruction{WalkLeft, WalkRight, WalkLeft}, row: 63},
-			{instructions: []WalkInstruction{WalkLeft, WalkRight, WalkRight}, row: 63},
-			{instructions: []WalkInstruction{WalkRight, WalkLeft, WalkLeft}, row: 63},
-			{instructions: []WalkInstruction{WalkRight, WalkLeft, WalkRight}, row: 63},
-			{instructions: []WalkInstruction{WalkRight, WalkRight, WalkLeft}, row: 63},
-			{instructions: []WalkInstruction{WalkRight, WalkRight, WalkRight, WalkLeft}, row: 63},
-			{instructions: []WalkInstruction{WalkRight, WalkRight, WalkRight, WalkRight, WalkLeft}, row: 63},
-			{instructions: []WalkInstruction{WalkRight, WalkRight, WalkRight, WalkRight, WalkRight, WalkLeft}, row: 63},
-		}},
+		{
+			"all paths for a basic 2x2", 2, 2, 2,
+			[]path{
+				{
+					row:          1,
+					instructions: []WalkInstruction{WalkLeft},
+				},
+				{
+					row:          1,
+					instructions: []WalkInstruction{WalkRight},
+				},
+			},
+			[]int{0, 1},
+		},
+		{
+			"all paths for a basic 4x4", 4, 2, 2,
+			[]path{
+				{
+					row:          0,
+					instructions: []WalkInstruction{WalkRight, WalkLeft},
+				},
+				{
+					row:          0,
+					instructions: []WalkInstruction{WalkRight, WalkRight},
+				},
+			},
+			[]int{0, 1},
+		},
+		{
+			"all paths for a basic 4x4 span more than 1 row", 4, 3, 2,
+			[]path{
+				{
+					row:          0,
+					instructions: []WalkInstruction{WalkRight, WalkRight},
+				},
+				{
+					row:          1,
+					instructions: []WalkInstruction{WalkLeft, WalkLeft},
+				},
+			},
+			[]int{0, 1},
+		},
+		{
+			"single share in the middle of a 128x128", 128, 8252, 1,
+			[]path{
+				{
+					row:          64,
+					instructions: []WalkInstruction{WalkLeft, WalkRight, WalkRight, WalkRight, WalkRight, WalkLeft, WalkLeft},
+				},
+			},
+			[]int{0},
+		},
+		{
+			"the 32nd path for the smallest blob with a subtree width of 128", 128, 0, 8193,
+			[]path{
+				{
+					row:          31,
+					instructions: []WalkInstruction{}, // there should be no instructions because we're using the first root
+				},
+			},
+			[]int{31},
+		},
+		{
+			"the 32nd path for the largest blob with a subtree width of 64", 128, 0, 8192,
+			[]path{
+				{
+					row:          15,
+					instructions: []WalkInstruction{WalkRight},
+				},
+			},
+			[]int{31},
+		},
+		{
+			"the 32nd path for the largest blob with a subtree width of 1", 128, 0, appconsts.SubtreeRootHeightThreshold,
+			[]path{
+				{
+					row:          0,
+					instructions: []WalkInstruction{WalkLeft, WalkLeft, WalkRight, WalkRight, WalkRight, WalkRight, WalkRight},
```

Comment:
[for reviewers]

Original Merkle tree spans 0-128
1. Left: 0 - 64
2. Left: 0 - 32
3. Right: 16-32
4. Right: 24-32
5. Right: 28-32
6. Right: 30-32
7. Right: 31-32